### ISSUE TITLE
High memory consumption in version 0.11.0 under stress.

### ISSUE THREAD

**Issue by arossert** [draios/sysdig#723]
There is a high memory usage when running sysdig under stress (open high volume of processes)

how to replicate:
- execute sysdig executable `sysdig > /dev/null`
- execute the stress command `while true; do date; done > /dev/null`
- stopped the stress for a while
- restart the stress 

(this is why we see the memory stopped raising and then raise again)

memory usage:
![memory](https://cloud.githubusercontent.com/assets/7085494/22069291/937f17c4-dda0-11e6-85c5-ee0e6ae1857f.png)

**Comment by mstemm** [comment]
I wasn't able to reproduce this with either 0.11.0 or 0.13.0 on ubuntu. I ran sysdig and the stress program as you described for 30 minutes and then watched the memory usage via top. In both cases, both the resident and virtual size stayed stable.

Can you provide more details on how you installed sysdig (source, binary package, container), the system on which you were running sysdig, and how you were recording memory usage?

**Comment by arossert** [comment]
I just tried this again with sysdig 0.13.0 on rhel5 and ubuntu 14.04.

I just downloaded the source from git and compile, in order to compile on rhel5 (kernel 2.6.18) there are some changes that need to be added:

In file driver/ppm_events.c (function compute_snaplen()):
```
// Use f_dentry for older kernel versions
#elif LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,20)
                if (file && file->f_dentry && file->f_dentry->d_inode) {
                        if (file->f_dentry->d_inode->i_rdev == PPM_NULL_RDEV) {
```

In file userspace/libsinsp/socket_headers.h:
```
#ifndef SOCK_NONBLOCK
#define SOCK_NONBLOCK 0
#endif
```

I just use a python script (psutils) to poll the memory usage (in %) and draw a graph.

This are the new results after about 1.5 hours:

Ubuntu:
![mem](https://cloud.githubusercontent.com/assets/7085494/22148474/f85b12e8-df16-11e6-8490-32ee366db73c.png)

rhel5:
![mem](https://cloud.githubusercontent.com/assets/7085494/22148483/019be5b2-df17-11e6-935a-63c28b1caa3c.png)

I can run some more tests if needed.

**Comment by mstemm** [comment]
To make sure I'm measuring exactly the same thing as you, can you tell me which psutils memory function you were using? I assume it was psutil.virtual_memory() and the percent column?

**Comment by mstemm** [comment]
Also could you try running the same test with the binary package of 0.13.0 and let us know if you see the same memory growth?

**Comment by mstemm** [comment]
I also built sysdig 0.13.0 from source on ubuntu 16.04 and ran sysdig and the date in a while loop as you described.

`sysdig-0.13.0/build$ sudo ./userspace/sysdig/sysdig  > /dev/null` and `$ while true; do date; done > /dev/null`

Here's the python script I ran using psutils:

```
import psutil
import time

proc = psutil.Process(10699)
while True:
    mem = proc.memory_percent()
    print mem
    time.sleep(5)
```

In my case, memory_percent has been stable at 0.742796410014.

Stumped as to why you're seeing memory growth.

**Comment by arossert** [comment]
We need the sysdig to work on RHEL5 (kernel 2.6.18), are the changes that i have made are correct?
(I have also committed a PR #695)

In file driver/ppm_events.c (function compute_snaplen()):
```
// Use f_dentry for older kernel versions
#elif LINUX_VERSION_CODE <= KERNEL_VERSION(2,6,20)
                if (file && file->f_dentry && file->f_dentry->d_inode) {
                        if (file->f_dentry->d_inode->i_rdev == PPM_NULL_RDEV) {
```
In file userspace/libsinsp/socket_headers.h:
```
#ifndef SOCK_NONBLOCK
#define SOCK_NONBLOCK 0
#endif
```

Maybe these changes cause the memory growth?
If these changes are correct, can you merge the PR?

**Comment by mstemm** [comment]
Looks like @luca3m is working with you on the related PR. I don't believe those changes would result in a memory leak. However, just to make sure can you try compiling the unmodified version on Ubuntu and see if you still see the memory growth?

**Comment by arossert** [comment]
I have done some tests on RHEL5 with the newest version 0.14.0 (with the compilation changes) and these are the results:

About 24 hours run, with this commands:
```sysdig proc.name!=sysdig > /dev/null```
```while true; do date; done```

![mem usage](https://cloud.githubusercontent.com/assets/7085494/22288230/a1ecc3fc-e2fe-11e6-968c-8cf83f4b1ee3.png)
![vms usage bytes](https://cloud.githubusercontent.com/assets/7085494/22288231/a1edb46a-e2fe-11e6-8e2a-3d2588454190.png)

These graphs represent the vms and usage(%).

It seems that the memory growth is very slow but its there, can it be from some STL containers (live vector and map) that double their size when more space is needed?

I will try to run the same test on Ubuntu and share the results

**Comment by arossert** [comment]
It seems that on Ubuntu we are not seeing this memory growth, I did the same test on Ubuntu server 14.04 and these are the results:
![ubuntu - mem usage](https://cloud.githubusercontent.com/assets/7085494/22324786/cfa3fdd2-e3b3-11e6-867e-74359cb0fc80.png)

![ubuntu - vms usage bytes](https://cloud.githubusercontent.com/assets/7085494/22324787/cfa7c4da-e3b3-11e6-932f-5cb6c62a0452.png)

Do you have any idea why this is happening on RHEL5 and not on Ubuntu?

**Comment by mstemm** [comment]
Not sure--I'd guess glibc is to blame. I'll try running the same test myself to see if I can figure out anything.

**Comment by mstemm** [comment]
Which specific RHEL5 version are you using? 5.11 with all updates?

**Comment by arossert** [comment]
I'm using RHEL5.4 with kernel updated to 2.6.18-409 to support tracepoints.

**Comment by mstemm** [comment]
Also, the gcc version that comes with RHEL 5.4 (4.1.2) is too old, so I guess you're using a different gcc version? Which version are you using and how did you install it (third-party yum repo, source, ?)

**Comment by arossert** [comment]
That's correct, I installed devtoolset-2 to get FCC 4.8.2. 
I think that you can get it from yum.

**Comment by mstemm** [comment]
After compiling 0.14.0 on RHEL 5.4 with all updates applied and using gcc 4.8.2 from devtoolset-2, I do see some slow memory growth. However, I can stop the memory growth by either disabling the code that writes formatted strings to standard out, or changing them to use direct fwrite() calls instead of `count << line`.

I first ran sysdig with the above workload for 10 minutes and measured `proc.memory_percent()` using the script above. memory_percent grew from 0.172918949585 to 0.243835148011 (for this VM, 3.5MB to 5.02MB RSS).

Next, I modified sysdig.cc as follows:

```
$ git diff -b userspace/sysdig/sysdig.cpp | cat
diff --git a/userspace/sysdig/sysdig.cpp b/userspace/sysdig/sysdig.cpp
index 149642c..9f15093 100644
--- a/userspace/sysdig/sysdig.cpp
+++ b/userspace/sysdig/sysdig.cpp
@@ -673,10 +673,12 @@ captureinfo do_inspect(sinsp* inspector,
                                        }
				}

-                               cout << line;
+                               //cout << line;
+                               fwrite(line.c_str(), line.size(), 1, stdout);
                                if(!json)
                                {
-                                       cout << endl;
+                                 //cout << endl;
+                                 fwrite("\n", 1, 1, stdout);
                                }
                        }
                }
```

With this change, I ran sysdig again for 10 minutes with the same workload. In this case, `proc.memory_percent()` stayed stable at 0.171364621949.

So I think this is a case of memory growth (maybe in c++'s cout?) rather than a memory leak. (Ideally I'd try direct leak checkers like valgrind but they don't handle all the system calls/ioctls we do with the driver right now).

I can't quite explain why there is no growth on ubuntu 14.04--it should have the same compiler version as RHEL 5.4 (gcc 4.8.2).

Also, the absolute growth is very modest--even in your day-long test, it was probably a change from ~3-4 MB to ~10-11 MB?

At this point I think we should just chalk it up to modest memory growth and consider the matter closed.
